### PR TITLE
Pin exact versions in Lisk Core Dockerfile

### LINKED ISSUE #714: Pin exact versions in Lisk Core Dockerfile
### Description

Improve the Dockerfile implementation as per the recommendations from the auditors

### Acceptance Criteria

- Build the image from the local codebase
- Use minimal base image
- Use proper environmental variables
- User is specified for execution commands
- Exact package version is used when using apk add commands

### PR DESCRIPTION
**Pull request by sameersubudhi**
### What was the problem?

This PR resolves #714 

### How was it solved?

- [x] Update Dockerfile to build image from the local codebase
- [x] Used minimal base image
- [x] User is specified for execution commands
- [x] Exact package version is used when using apk add commands 

### How was it tested?

- Local both on MacOS and Linux
  - `docker system prune`
  - `docker build -f docker/Dockerfile -t lisk/core .`
  - `docker run --rm -it lisk/core start -n devnet --api-ws -l info`

- To build using the `Makefile`:
  - `cd docker`
  - `make build`
  - `docker run --rm -it lisk/core:4.0.0-alpha.12 start -n devnet --api-ws -l info`

### CODE REVIEW

**Review by nagdahimanshu — APPROVED** [review]
Looks good and works as expected 👍

File: docker/Dockerfile [hunk 1]
```diff
@@ -1,33 +1,37 @@
-ARG NODEJS_VERSION=16
-FROM node:$NODEJS_VERSION-alpine
+##### Stage 1
 
-ARG CORE_VERSION
-ARG REGISTRY_URL
-ARG REGISTRY_AUTH_TOKEN
+FROM node:16-alpine AS builder
```
Comment (ManuGowda) [inline docker/Dockerfile:3]: Its good to use `ARG`? this way run time you can change to whatever version we prefer?

File: docker/Dockerfile [hunk 2]
```diff
@@ -1,33 +1,37 @@
...
-ENV NODE_ENV=production
+RUN apk add --no-cache alpine-sdk=1.0-r1 python3=3.10.10-r0 libtool=2.4.7-r1 autoconf=2.71-r1 automake=1.16.5-r1 rust=1.64.0-r2 cargo=1.64.0-r2 cmake=3.24.3-r0 clang15-libs=15.0.7-r0 clang15-dev=15.0.7-r0 clang15=15.0.7-r0 rustfmt=1.64.0-r2 linux-headers=5.19.5-r0 && \
+    adduser -D builder && \
+    mkdir /home/builder/build && \
+    chown -R builder:builder /home/builder/ && \
+    npm install -g npm
 
-RUN apk --no-cache upgrade && \
-    apk --no-cache add --virtual build-dependencies alpine-sdk autoconf automake libtool linux-headers python2
+USER builder
```
Comment (ManuGowda) [inline docker/Dockerfile:11]: why `builder` as user? we used `lisk` as the user before?

File: docker/Dockerfile [hunk 3]
```diff
@@ -1,33 +1,37 @@
...
-WORKDIR /home/lisk
 
-RUN if [ -n "$REGISTRY_URL" ]; then \
-      echo -e "registry=$REGISTRY_URL/\n${REGISTRY_URL#*:}/:_authToken=$REGISTRY_AUTH_TOKEN" >/home/lisk/.npmrc; \
-    fi
-RUN npm install lisk-core@$CORE_VERSION && \
-    rm -f /home/lisk/.npmrc
+##### Stage 2
+
+FROM node:16-alpine
```
Comment (ManuGowda) [inline docker/Dockerfile:20]: if we use the ARG, we could reuse the same here?

**Review by Nazgolze — APPROVED** [review]

**Review by nagdahimanshu — APPROVED** [review]

**Review by ManuGowda — APPROVED** [review]

**Pull request merged**